torch_disc/torch_disc/csrc/disc_jit.cpp

```diff
@@ -161,29 +147,95 @@ std::string MhloConversaion(const std::shared_ptr<Graph>& graph) {
   std::ofstream outfile(in_fname);
   outfile << parsable_mlir << std::endl;
   outfile.flush();
-  return in_fname;
+  return std::make_tuple(in_fname, input_dev_str, output_dev_str);
 }
 
-void CallDiscCompiler(const std::string& mlir_fname) {
+std::string CallDiscCompiler(const std::string& mlir_fname) {
   std::string out_fname = mlir_fname + ".out";
   std::string cmd = DiscCMD(mlir_fname, out_fname);
   TORCH_CHECK(std::system(cmd.c_str()) == 0,
               "disc compile failed with cmd: " + cmd);
+  return out_fname;
 }
 
-void DiscCompilation(const std::shared_ptr<Graph>& graph) {
-  for (auto node : graph->nodes()) {
-    if (node->kind() == prim::FusionGroup) {
-      auto sub_graph = node->g(attr::Subgraph);
-      std::string fname = MhloConversaion(sub_graph);
-      CallDiscCompiler(fname);
+std::vector<const torch::jit::Value*> ToConstValue(
+    c10::ArrayRef<torch::jit::Value*> values) {
+  std::vector<const torch::jit::Value*> const_vals;
+  std::copy(values.begin(), values.end(), std::back_inserter(const_vals));
+  return const_vals;
+}
+
+std::vector<c10::IValue> DiscCompilation(const std::shared_ptr<Graph>& graph) {
+  std::vector<c10::IValue> disc_inputs;
+  std::vector<torch::jit::Node*> disc_nodes;
+  std::copy_if(graph->nodes().begin(), graph->nodes().end(),
+               std::back_inserter(disc_nodes), [](torch::jit::Node* node) {
+                 return node->kind() == prim::FusionGroup;
+               });
+
+  for (auto node : disc_nodes) {
+    auto sub_graph = node->g(attr::Subgraph);
+
+    auto option = DiscClass::MakeOption();
+    std::string mlir_fname;
+    std::tie(mlir_fname, option->input_dev_str, option->output_dev_str) =
+        MhloConversaion(sub_graph);
+    auto output_fname = CallDiscCompiler(mlir_fname);
+    option->executable_prog_bytes = ReadFileBytes(output_fname);
+    option->constant_bytes = ReadFileBytes(output_fname + ".pbtxt");
+    option->input_type_spec_str =
+        torch::blade::ShapeTypeSpec::GetShapeTypeSpec(
+            ToConstValue(sub_graph->inputs()), false /**force_concrete**/)
+            .Serialize();
+    option->output_type_spec_str =
+        torch::blade::ShapeTypeSpec::GetShapeTypeSpec(
+            ToConstValue(sub_graph->outputs()), false /**force_concret**/)
+            .Serialize();
+
+    // add DiscClass object as graph input
+    c10::IValue disc_val = torch::make_custom_class<DiscClass>(option);
```

Comment:
Nit: Extract L196-L228 into a standalone function, since they work together to modify the graph.

Reply:
done.